utils/utils.py

```diff
@@ -182,3 +184,19 @@ def update_query_params(url: str, **params: str) -> str:
 
 def as_p_tags(texts: Iterable[str]) -> str:
     return "".join(f"<p>{p}</p>" for p in texts)
+
+
+def convert_html_to_text(html_text: str) -> str:
+    text = html2text(html_text, bodywidth=0)
+
+    # Link text and url are the same
+    # Remove angle-brackets from links `<url>` -> `url`
+    # If there is a dot after the link, add a space between the link and the dot.
+    text = re.sub(r"<(https?://[^>]+)>(\.?)", r"\1 \2", text)
+
+    # Link text and url are different
+    # Replace markdown-style links `[text](url)` with `text <url>`
+    text = re.sub(r"\[([^\]]+)\]\(((http|https)?://[^\)]+)\)", r"\1 <\2>", text)
```

Comment:
For complex regexes like this, I would format them manually, and use named capture groups for the groups I care about. here is my take (with small improvements I think)

```python
# fmt: off
x = (
    r"\["                            # begins with "["
    r"(?P<text>[^\]]+)"              # any text that is not "]"
    r"\]"                            # followed with "]"
    r"\("                            # followed by "("
    r"(?P<link>(https?://)?[^\)]+)"  # any link, with optional protocol
    r"\)"                            # followed by ")"
)
# fmt: on

text = re.sub(x, r"\g<text> <\g<link>>", text)
```